### PR TITLE
Add `fn:substring` to Metapath function library

### PR DESCRIPTION
**Pull request by aj-stein-gsa**
# Committer Notes

This PR adds `fn:substring`, 1 of 3 to be added as part of work for #132.

### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/metaschema-framework/metaschema-java/blob/main/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/metaschema-framework/metaschema-java/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you included examples of how to use your new feature(s)?
- ~Have you updated all website and readme documentation affected by the changes you made?~

### CODE REVIEW

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Implements <a href=
+ * "https://www.w3.org/TR/xpath-functions-30/#func-substring">fn:substring</a>.
```
Comment (david-waltermire) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java:29]:
```suggestion
 * "https://www.w3.org/TR/xpath-functions-31/#func-substring">fn:substring</a>.
```

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      .focusIndependent()
+      .argument(IArgument.builder()
+          .name("sourceString")
+          .type(IStringItem.class)
+          .zeroOrOne()
+          .build())
+      .argument(IArgument.builder()
+          .name("start")
+          // Review if xs:double to IIntegerItem data-type mapping appropriate?
+          .type(IIntegerItem.class)
```
Comment (david-waltermire) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java:47]:
```suggestion
          .type(IDecimalItem.class)
```

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+          .zeroOrOne()
+          .build())
+      .argument(IArgument.builder()
+          .name("length")
+          // Review if xs:double to IIntegerItem data-type mapping appropriate?
+          .type(IIntegerItem.class)
```
Comment (david-waltermire) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java:53]:
```suggestion
          .type(IDecimalItem.class)
```

Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java:53]: Need to fix other occurrences of this argument below.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+          .zeroOrOne()
```
Comment (david-waltermire) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java:54]:
```suggestion
          .one()
```

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+      .argument(IArgument.builder()
+          .name("sourceString")
+          .type(IStringItem.class)
+          .zeroOrOne()
+          .build())
+      .argument(IArgument.builder()
+          .name("start")
+          // Review if xs:double to IIntegerItem data-type mapping appropriate?
+          .type(IIntegerItem.class)
+          .zeroOrOne()
```
Comment (david-waltermire) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java:48]:
```suggestion
          .one()
```

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+          .build())
+      .argument(IArgument.builder()
+          .name("length")
+          // Review if xs:double to IIntegerItem data-type mapping appropriate?
+          .type(IIntegerItem.class)
+          .zeroOrOne()
+          .build())
+      .allowUnboundedArity(true)
```
Comment (david-waltermire) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java:56]:
```suggestion
```

File: core/src/test/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstringTest.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+class FnSubstringTest
+    extends ExpressionTestBase {
+  private static Stream<Arguments> provideValues() { // NOPMD - false positive
+    return Stream.of(
+        Arguments.of(
+            ISequence.of(string("ada")),
+            "substring('metadata', 4, 3)"));
+  }
```
Comment (david-waltermire) [inline core/src/test/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstringTest.java:30]: Do you plan to expand this to represent all the examples?

**Comment by aj-stein-gsa** [comment]
Hey @david-waltermire! I know I have a lot to do and the integer data types and lazy implementation was because I needed time to think
more about the decimal/double handling for the edge cases that weren't integer based.

I'll handle your very welcome feedback soon enough!

File: core/src/test/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstringTest.java (hunk 7 above)
Comment (aj-stein-gsa) [reply core/src/test/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstringTest.java:30]: Oh absolutely I left this in draft because much left to do but wanted one passing test when I pushed up changes last night. Rest assured I will add them all!

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/DefaultFunctionLibrary.java [hunk 8]
```diff
@@ -146,10 +146,11 @@ public DefaultFunctionLibrary() { // NOPMD - intentional
     // https://www.w3.org/TR/xpath-functions-31/#func-static-base-uri
     registerFunction(FnStaticBaseUri.SIGNATURE);
     // P0: https://www.w3.org/TR/xpath-functions-31/#func-string
+    registerFunction(FnSubstring.SIGNATURE_THREE_ARG);
```
Comment (aj-stein-gsa) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/DefaultFunctionLibrary.java:149]: @david-waltermire, having reviewed your other PR, for optional arguments I should be expressly declaring different arity functions separately? I don't remember doing that before and somehow dropping optional arguments in code, so I assume I should be adding this function registry change?

```suggestion
    registerFunction(FnSubstring.SIGNATURE_TWO_ARG);
    registerFunction(FnSubstring.SIGNATURE_THREE_ARG);
```

If so I'll adjust the other parts of the code, but I had not noticed this pattern before, so I thought I'd just ask.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java (hunk 6 above)
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java:56]: Unbounded arity means the last argument can repeat multiple times.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/DefaultFunctionLibrary.java (hunk 8 above)
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/DefaultFunctionLibrary.java:149]: Specifying each variant is important to support the type coercion needed to support the function call. This happens automatically by the code that calls the function.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/DefaultFunctionLibrary.java [hunk 9]
```diff
@@ -152,10 +152,13 @@ public DefaultFunctionLibrary() { // NOPMD - intentional
     // https://www.w3.org/TR/xpath-functions-31/#func-string
     registerFunction(FnString.SIGNATURE_NO_ARG);
     registerFunction(FnString.SIGNATURE_ONE_ARG);
+    // P0: https://www.w3.org/TR/xpath-functions-31/#func-string
+    registerFunction(FnSubstring.SIGNATURE_TWO_ARG);
+    registerFunction(FnSubstring.SIGNATURE_THREE_ARG);
```
Comment (david-waltermire) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/DefaultFunctionLibrary.java:157]: This appears to be in the wrong place
```suggestion
```

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/DefaultFunctionLibrary.java [hunk 10]
```diff
@@ -152,10 +152,13 @@ public DefaultFunctionLibrary() { // NOPMD - intentional
...
     // P1: https://www.w3.org/TR/xpath-functions-31/#func-string-join
     // P1: https://www.w3.org/TR/xpath-functions-31/#func-string-length
     // P1: https://www.w3.org/TR/xpath-functions-31/#func-subsequence
-    // P1: https://www.w3.org/TR/xpath-functions-31/#func-substring
+    // https://www.w3.org/TR/xpath-functions-31/#func-substring
```
Comment (david-waltermire) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/DefaultFunctionLibrary.java:161]:
```suggestion
    // https://www.w3.org/TR/xpath-functions-31/#func-substring
    registerFunction(FnSubstring.SIGNATURE_TWO_ARG);
    registerFunction(FnSubstring.SIGNATURE_THREE_ARG);

```

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+   * @return the atomized result
+   */
+  @NonNull
+  public static IStringItem fnSubstring(@NonNull IStringItem sourceString, @NonNull IDecimalItem start,
+      @NonNull IDecimalItem length) {
+
+    int startIndex = start.round().asInteger().intValue();
+    int endIndex = length.round().asInteger().intValue();
+    return fnSubstring(sourceString, startIndex, endIndex);
+  }
```
Comment (david-waltermire) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java:149]: Is this version used?

Comment (aj-stein-gsa) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java]: We got rid of this during our last pairing session.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/DefaultFunctionLibrary.java (hunk 10 above)
Comment (aj-stein-gsa) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/DefaultFunctionLibrary.java:164]: It seems we have relocated it in the interim.

**Comment by aj-stein-gsa** [comment]
So either there is some configuration issue with Javadocs or maybe I am now in charge of writing a lot of docs for those functions, per the build errors?

**Comment by aj-stein-gsa** [comment]
> So either there is some configuration issue with Javadocs or maybe I am now in charge of writing a lot of docs for those functions, per the build errors?

So, @david-waltermire, obviously the function signature slip ups were on me, but how did only updating the Javadoc plugin for Maven fix those issues? Or does not compiling the function correctly in one place trickle down code gen problems so all related docs fail to build?

**Comment by david-waltermire** [comment]
> > So either there is some configuration issue with Javadocs or maybe I am now in charge of writing a lot of docs for those functions, per the build errors?
> 
> So, @david-waltermire, obviously the function signature slip ups were on me, but how did only updating the Javadoc plugin for Maven fix those issues? Or does not compiling the function correctly in one place trickle down code gen problems so all related docs fail to build?

Updating the plugin didn't. I did that thinking it was a Javadoc issue and just kept the bump. The issue was a Javadoc error caused by a javadoc parameter entry that didn't exist.

**Review by david-waltermire — APPROVED** [review]
